File: LightOJ/1153.py
from collections import deque

def bfs(capacity, source, sink, parent):
    queue = deque([source])
    visited = set([source])
    while queue:
        u = queue.popleft()
        for v in range(len(capacity)):
            if v not in visited and capacity[u][v] > 0:
                queue.append(v)
                visited.add(v)
                parent[v] = u
                if v == sink:
                    return True
    return False

def edmonds_karp(capacity, source, sink):
    max_flow = 0
    n = len(capacity)
    parent = [-1] * n

    while bfs(capacity, source, sink, parent):
        path_flow = float('inf')
        v = sink
        while v != source:
            u = parent[v]
            path_flow = min(path_flow, capacity[u][v])
            v = u

        v = sink
        while v != source:
            u = parent[v]
            capacity[u][v] -= path_flow
            capacity[v][u] += path_flow
            v = u

        max_flow += path_flow

    return max_flow


T = int(input().strip())

for case in range(1, T+1):
    n = int(input().strip())
    s, t, c = map(int, input().split())
    
    capacity = [[0] * n for _ in range(n)]
    
    for _ in range(c):
        u, v, bw = map(int, input().split())
        u -= 1
        v -= 1
        capacity[u][v] += bw

File: LightOJ/1153.cpp
#include <bits/stdc++.h>
using namespace std;
const int INF = numeric_limits<int>::max();

int bfs(int s, int t,
        vector<int>& parent,
        vector<vector<int>>& capacity,
        vector<vector<int>>& adj){
    fill(parent.begin(), parent.end(), -1);
    parent[s] = -2;
    queue<pair<int, int>> q;
    q.push({s, INF});

    while (!q.empty()) {
        int cur = q.front().first;
        int flow = q.front().second;
        q.pop();

        for (int next : adj[cur]) {
            if (parent[next] == -1 && capacity[cur][next]) {
                parent[next] = cur;
                int new_flow = min(flow, capacity[cur][next]);
                if (next == t)
                    return new_flow;
                q.push({next, new_flow});
            }
        }
    }

    return 0;
}

int maxflow(int n, int s, int t,vector<vector<int>>& capacity,vector<vector<int>>& adj ){
    int flow = 0;
    vector<int> parent(n);
    int new_flow;

    while ((new_flow = bfs(s, t, parent, capacity, adj))) {
        flow+=new_flow;
        int cur=t;
        while(cur != s){
            int prev = parent[cur];
            capacity[prev][cur] -=new_flow;
            capacity[cur][prev]+=new_flow;
            cur=prev;
        }
    }
    return flow;
}

int main(){
    int T;
    cin >> T;
    int i = 1;
    while(T--){
        int n;
        cin >> n;
        int s; int t; int c;
        cin >> s >> t >> c;
        s--;
        t--;
        vector<vector<int>> capacity(n, vector<int>(n, 0));
        vector<vector<int>> adj(n);
        while(c--){
            int u; int v; int bw; 
            cin >> u >> v >> bw;
            u--;
            v--;
            capacity[u][v] += bw;
            capacity[v][u] += bw;
            adj[u].push_back(v);
            adj[v].push_back(u);
        }

        int flow = maxflow(n,s,t,capacity, adj);
        cout << "Case " << i << ": " << flow << endl;
        i++;

    }


    return 0;
}